src/kernel-install/ukify.py

```diff
@@ -0,0 +1,233 @@
+#!/usr/bin/python
+# SPDX-License-Identifier: LGPL-2.1+
+
+import argparse
+import dataclasses
+import itertools
+import pathlib
+import os
+import shlex
+import subprocess
+import tempfile
+import typing
+
+efi_arches = {
+        # host_arch: efi_arch
+        'x86'    : 'ia32',
+        'x86_64' : 'x64',
+        'arm'    : 'arm',
+        'aarch64': 'aa64',
+        'riscv64': 'riscv64',
+}
+
+def guess_efi_arch():
+    arch = os.uname().machine
+    efi_arch = efi_arches[arch]
+
+    if arch == 'x86_64':
+        try:
+            size = open('/sys/firmware/efi/fw_platform_size').read().strip()
+        except FileNotFoundError:
+            pass
+        else:
+            if int(size) == 32:
+                efi_arch = 'ia32'
+
+    print(f'Host arch {arch!r}, efi arch {efi_arch!r}')
+    return efi_arch
+
+def shell_join(cmd):
+    return ' '.join(shlex.quote(str(x)) for x in cmd)
+
+@dataclasses.dataclass
+class Section:
+    name: str
+    content: str
+    offset: int
+    tmpfile: typing.Optional[typing.IO] = None
+
+    @classmethod
+    def create(cls, name, content, offset):
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            content = tmp.name
+        else:
+            tmp = None
+
+        return cls(name, content, offset, tmpfile=tmp)
+
+    @classmethod
+    def parse_arg(cls, s):
+        parts = s.split(':')
+        if len(parts) < 2 or len(parts) > 3:
+            raise ValueError(f'Cannot parse section spec: {s!r}')
+        name, content = parts[:2]
+        offset = int(parts[2], base=0) if len(parts) >= 3 else None
+        return cls.create(name, content, offset)
+
+
+def parse_args():
+    p = argparse.ArgumentParser(
+        description='Build and sign Unified Kernel Images',
+        usage='''\
+usage: ukify [options…] linux initrd
+       ukify -h | --help
+''')
+
+    # Suppress printing of usage synopsis on errors
+    p.error = lambda message: p.exit(2, f'{p.prog}: error: {message}\n')
+
+    p.add_argument('linux',
+                   type=pathlib.Path)
+    p.add_argument('initrd',
+                   type=pathlib.Path)
+
+    p.add_argument('--cmdline')
+
+    p.add_argument('--os-release',
+                   type=pathlib.Path)
+
+    p.add_argument('--devicetree',
+                   type=pathlib.Path)
+    p.add_argument('--splash',
```

Comment:
I added a call to `PIL.Image.open()` to verify that the image looks valid. If we should have some check on dimensions or such, this can be easily extended.